Engine/Utilities.cs

```diff
@@ -21,177 +22,259 @@
             public Vector3f n;
             public Vector3f tx;
+        };
+        public List<PrimitiveVertex> vertices = null;
+        public List<int> indices = null;
+
+        private Buffer vertexBuffer;
+        private VertexBufferBinding vertexBufferBinding;
+        private Buffer indexBuffer;
+
+        // material assigned on mesh load
+        public int defaultMaterialIndex;
+
+        ~Primitive()
+        {
+            Dispose(disposing: false);
         }
-        private struct ModelVertexWithIndex
+
+        public void GenerateGPUData()
         {
-            public ModelVertex v;
-            public uint index;
-            public ModelVertexWithIndex(ModelVertex v, uint index)
-            {
-                this.v = v;
-                this.index = index;
-            }
-            public override bool Equals(object obj)
-            {
-                if (obj is ModelVertexWithIndex)
-                    return v.Equals(((ModelVertexWithIndex)obj).v);
-                return false;
-            }
-            public override int GetHashCode()
+            if (vertices == null || indices == null)
+                throw new Exception("Geometry data can't be empty.");
+
+            for (int i = 0; i < indices.Count / 3; i++)
             {
-                return v.GetHashCode();
+                Vector3f edge1 = vertices[i * 3 + 1].v - vertices[i * 3 + 0].v;
+                Vector3f edge2 = vertices[i * 3 + 2].v - vertices[i * 3 + 0].v;
+                Vector2f UVedge1 = vertices[i * 3 + 1].t - vertices[i * 3 + 0].t;
+                Vector2f UVedge2 = vertices[i * 3 + 2].t - vertices[i * 3 + 0].t;
+                Vector3f tx = ((edge1 * UVedge2.y - edge2 * UVedge1.y) / (UVedge1.x * UVedge2.y - UVedge1.y * UVedge2.x)).normalized();
+                PrimitiveVertex vertex0 = vertices[i * 3 + 0];
+                PrimitiveVertex vertex1 = vertices[i * 3 + 1];
+                PrimitiveVertex vertex2 = vertices[i * 3 + 2];
+
+                vertex0.tx = new Vector3f(tx.x, tx.y, tx.z);
+                vertex1.tx = new Vector3f(tx.x, tx.y, tx.z);
+                vertex2.tx = new Vector3f(tx.x, tx.y, tx.z);
+
+                vertices[i * 3 + 0] = vertex0;
+                vertices[i * 3 + 1] = vertex1;
+                vertices[i * 3 + 2] = vertex2;
             }
+
+            vertexBuffer = Buffer.Create(GraphicsCore.CurrentDevice, BindFlags.VertexBuffer, vertices.ToArray());
+            vertexBufferBinding = new VertexBufferBinding(vertexBuffer, Utilities.SizeOf<PrimitiveVertex>(), 0);
+            indexBuffer = Buffer.Create(GraphicsCore.CurrentDevice, BindFlags.IndexBuffer, indices.ToArray());
         }
 
-        public void updateModel()
+        public void Render()
         {
-            if (v == null || v_i == null)
-                throw new Exception("Model can't be empty.");
+            if (disposed)
+                throw new Exception("Trying to render disposed primitive");
```

Comment:
there is a special error type for disposed object, refer to any other resource implementing IDisposable interface here for code example